# **Laboratorio 03**

In [1]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/netflix_titles.csv')
df.head()

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


**Parte 1: Limpieza y preparación**

In [2]:
#Número de filas y columnas.

df.shape

(8807, 12)

In [3]:
#Tipos de datos.

df.dtypes

,0
show_id,object
type,object
title,object
director,object
cast,object
country,object
date_added,object
release_year,int64
rating,object
duration,object


In [4]:
#Valores nulos por columna.

df.isnull().sum()

,0
show_id,0
type,0
title,0
director,2634
cast,825
country,831
date_added,10
release_year,0
rating,4
duration,3


In [6]:
#Convertir date_added a tipo fecha

df["date_added"] = df["date_added"].str.strip()
df["date_added"] = pd.to_datetime(df["date_added"])

In [7]:
#Crear columnas con assign

df = df.assign(
    year_added=df["date_added"].dt.year,
    month_added=df["date_added"].dt.month
)

**Parte 2: Técnicas Avanzadas de pandas**

In [8]:
#.loc

df_movies_2018 = df.loc[
    (df["type"] == "Movie") & (df["year_added"] > 2018)
]

In [9]:
#str.contains() --> Títulos con "love"

df_love = df.loc[
    df["title"].str.contains("love", case=False, na=False)
]

#str.extract() --> duración (min)

df["duration_minutes"] = df["duration"].str.extract(r"(\d+)").astype(float)

In [10]:
#explode()

df_exploded = df.explode("listed_in")

In [11]:
#Top 10 géneros más frecuentes

top_genres = df_exploded["listed_in"].value_counts().head(10)

In [13]:
#where() y mask()

df["is_long_where"] = df["duration_minutes"].where(
    df["duration_minutes"] > 120,
    other="No largo"
)

df["is_long_mask"] = df["duration_minutes"].mask(
    df["duration_minutes"] <= 120,
    other="No largo"
)

In [14]:
#.loc

df_filtered = df.loc[
    (df["duration_minutes"] > 100) &
    (df["rating"] == "R") &
    (df["country"] == "United States")
]

In [15]:
#Top 10 películas más largas



,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description,year_added,month_added,duration_minutes,is_long_where,is_long_mask
4253,s4254,Movie,Black Mirror: Bandersnatch,nan,"Fionn Whitehead, Will Poulter, Craig Parkinson, Alice Lowe, Asim Chaudhry",United States,2018-12-28 00:00:00,2018,TV-MA,312 min,"Dramas, International Movies, Sci-Fi & Fantasy","In 1984, a young programmer begins to question reality as he adapts a dark fantasy novel into a video game. A mind-bending tale with multiple endings.",2018.000000,12.000000,312,312.000000,312.000000
717,s718,Movie,Headspace: Unwind Your Mind,nan,"Andy Puddicombe, Evelyn Lewis Prieto, Ginger Daniels, Darren Pettie, Simon Prebble, Rhiannon Mcgavin, Kate Seftel",nan,2021-06-15 00:00:00,2021,TV-G,273 min,Documentaries,"Do you want to relax, meditate or sleep deeply? Personalize the experience according to your mood or mindset with this Headspace interactive special.",2021.000000,6.000000,273,273.000000,273.000000
2491,s2492,Movie,The School of Mischief,Houssam El-Din Mustafa,"Suhair El-Babili, Adel Emam, Saeed Saleh, Younes Shalabi, Hadi El-Gayyar, Ahmad Zaki, Hassan Moustafa",Egypt,2020-05-21 00:00:00,1973,TV-14,253 min,"Comedies, Dramas, International Movies",A high school teacher volunteers to transform five notorious misfits into model students — and has unintended results.,2020.000000,5.000000,253,253.000000,253.000000
2487,s2488,Movie,No Longer kids,Samir Al Asfory,"Said Saleh, Hassan Moustafa, Ahmed Zaki, Younes Shalabi, Nadia Shukri, Karima Mokhtar",Egypt,2020-05-21 00:00:00,1979,TV-14,237 min,"Comedies, Dramas, International Movies","Hoping to prevent their father from skipping town with his mistress, four rowdy siblings resort to absurd measures to stop him.",2020.000000,5.000000,237,237.000000,237.000000
2484,s2485,Movie,Lock Your Girls In,Fouad El-Mohandes,"Fouad El-Mohandes, Sanaa Younes, Sherihan, Ahmed Rateb, Ijlal Zaki, Zakariya Mowafi",nan,2020-05-21 00:00:00,1982,TV-PG,233 min,"Comedies, International Movies, Romantic Movies",A widower believes he must marry off his three problematic daughters before he can pursue his real goal of marrying his secret love.,2020.000000,5.000000,233,233.000000,233.000000
2488,s2489,Movie,Raya and Sakina,Hussein Kamal,"Suhair El-Babili, Shadia, Abdel Moneim Madbouly, Ahmed Bedir",nan,2020-05-21 00:00:00,1984,TV-14,230 min,"Comedies, Dramas, International Movies","When robberies and murders targeting women sweep early 20th-century Egypt, the hunt for suspects leads to two shadowy sisters. Based on a true story.",2020.000000,5.000000,230,230.000000,230.000000
166,s167,Movie,Once Upon a Time in America,Sergio Leone,"Robert De Niro, James Woods, Elizabeth McGovern, Treat Williams, Tuesday Weld, Burt Young, Joe Pesci, Danny Aiello, William Forsythe, James Hayden","Italy, United States",2021-09-01 00:00:00,1984,R,229 min,"Classic Movies, Dramas",Director Sergio Leone's sprawling crime epic follows a group of Jewish mobsters who rise in the ranks of organized crime in 1920s New York City.,2021.000000,9.000000,229,229.000000,229.000000
7932,s7933,Movie,Sangam,Raj Kapoor,"Raj Kapoor, Vyjayanthimala, Rajendra Kumar, Lalita Pawar, Achala Sachdev, Hari Shivdasani, Raj Mehra, Iftekhar",India,2019-12-31 00:00:00,1964,TV-14,228 min,"Classic Movies, Dramas, International Movies","Returning home from war after being assumed dead, a pilot weds the woman he has long loved, unaware that she had been planning to marry his best friend.",2019.000000,12.000000,228,228.000000,228.000000
1019,s1020,Movie,Lagaan,Ashutosh Gowariker,"Aamir Khan, Gracy Singh, Rachel Shelley, Paul Blackthorne, Kulbhushan Kharbanda, Raghuvir Yadav, Yashpal Sharma, Rajendranath Zutshi, Rajesh Vivek, Aditya Lakhia","India, United Kingdom",2021-04-17 00:00:00,2001,PG,224 min,"Dramas, International Movies, Music & Musicals","In 1890s India, an arrogant British commander challenges the harshly taxed residents of Champaner to a high-stakes crick

**Pregunta Desafío**



In [18]:
df_genre_rating = df.explode("listed_in")

df_genre_rating[["listed_in", "rating"]].value_counts().head(10)

,,count
listed_in,rating,
Stand-Up Comedy,TV-MA,284
"Dramas, International Movies",TV-MA,154
"Dramas, Independent Movies, International Movies",TV-MA,142
"Dramas, International Movies",TV-14,139
"Comedies, Dramas, International Movies",TV-14,139
Kids' TV,TV-Y,117
Documentaries,TV-MA,115
"Children & Family Movies, Comedies",PG,99
Documentaries,TV-14,98


**¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?**


Los resultados muestran que las combinaciones más frecuentes entre género y rating se concentran principalmente en contenido de tipo “Stand-Up Comedy” con clasificación “TV-MA”, siendo la combinación más repetida del dataset (284 apariciones).

En segundo lugar, destacan combinaciones de dramas e internacional movies, especialmente con ratings TV-MA y TV-14, lo que indica una fuerte presencia de contenido adulto o para audiencias mayores en producciones internacionales.

También aparecen con frecuencia combinaciones mixtas como:

Dramas + Independent + International Movies (TV-MA)
Comedias + Dramas + Internacionales (TV-14)

Esto sugiere que el catálogo de Netflix no está dominado por géneros “puros”, sino por contenidos híbridos, donde los dramas y las producciones internacionales tienen un rol central.

En categorías familiares o infantiles, aparecen combinaciones como:

Kids’ TV (TV-Y)
Children & Family Movies (PG)

pero con menor frecuencia en comparación con contenido adulto.

**¿Existen películas con el mismo nombre (tittle) pero con distinto año de lanzamiento (release_year)?**

No, no existen películas con el mismo nombre (title) que presenten distintos años de lanzamiento (release_year) en el dataset. Esto indica que, aunque puedan existir títulos repetidos, estos corresponden al mismo año de lanzamiento y no a versiones diferentes del contenido en años distintos.

In [23]:
df.loc[
    df.duplicated(subset=["title"], keep=False),
    ["title", "release_year"]
].duplicated(subset=["title", "release_year"]).any()

np.False_

**¿Cuántos títulos únicos hay en total en la columna title?**

En el dataset existen 8807 títulos únicos en la columna title. Esto indica que cada registro corresponde a un contenido distinto, sin repeticiones de nombres en la base de datos.

In [24]:
df["title"].nunique()

8807